postprocessing/science/spectral_corner_frequency_from_surface_xdmf.py

```diff
@@ -0,0 +1,382 @@
+    amplitude_tensor[:,0,0] = amplitude_tr
+    
+    # Initialise tensor of possible Q_zero values depending on amplitude_tr[0] or take mean amplitude below corner frequency
+    if q_dim != 0:
+        q_zero = np.array([[np.geomspace(0.1*amplitude_tr[0], 10*amplitude_tr[0], num=q_dim)]])
+    else:
+        q_zero = np.empty((1,freq_tr[:fc_ind].shape[0], 1))
+        for i in range(q_zero.shape[1]):
+            q_zero[0,i,0] = np.mean(amplitude_tr[:i+1])
+    
+    # Calculate Brune spectra of every (Q, fc) combination by broadcasting
+    brune_function = q_zero*(1/(1+(freq_tensor/fc_tensor)**n))
+    
+    # Calculate the misfit of every (Q, fc) combination
+    if misfit == "RMS":
+        misfit_tensor = (brune_function - amplitude_tensor)**2
+        misfit_tensor = np.sqrt(np.mean(misfit_tensor, axis=0))
+    if misfit == "SSM":
+        misfit_tensor = np.abs(np.log10(amplitude_tensor/brune_function))
+        misfit_tensor = np.mean(misfit_tensor, axis=0)
+    
+    # Get indices of minimum misfit (Q, fc) combination
+    index = np.unravel_index(np.argmin(misfit_tensor), misfit_tensor.shape)
+    
+    fc = fc_tensor[0,index[0],0]
+    if q_dim != 0:
+        q = q_zero[0,0,index[1]]
+    else:
+        q = q_zero[0,index[1],0]
+        
+    return fc, q
+
+def complete_fft(tr, dt):
+    tr_fft=np.fft.rfft(tr.data)
+    times_comp = np.fft.rfftfreq(tr.shape[0],dt)
+    return tr_fft, times_comp
+
+def ApproximateEventDurationAndHypocenter(slipRateThreshold=args.slipRateThreshold[0]):
+    # Calculates the derivative of ASl (probably faster than loading SR1 and SR2)
+    # and approximates event duration by multiplying dt with the number of timesteps,
+    # where maximum on-fault slip rate is above a threshold (slipRateThreshold)
+    if nprocs == 1 or args.serialLoading:
+        ASl = faultxdmf.ReadData("ASl").T[::stepsize,timeIndicesFault[0]:timeIndicesFault[1]]
+    else:
+        ASl = LoadSingleComponentParallel("ASl", surfaceData=False)
+    duration = np.sum(np.where(np.amax((ASl[:,1:]-ASl[:,:-1])/dtFault, axis=0) > slipRateThreshold, 1, 0))*dtFault
+    slippingFault = np.where((ASl[:,1:]-ASl[:,:-1])/dtFault > slipRateThreshold, 1, 0)
+    indHypocenter = np.argmax(slippingFault[:,np.argmax(np.amax(slippingFault, axis=0))])
+    hypocenter = faultxyz[indHypocenter,:]
+    return duration, hypocenter
+
+def PickPPhases(trigger=0.001):                 # This function is only suited for synthetic data without noise
+    data = np.sqrt(u**2 + v**2 + w**2)
+    thresholds = np.amax(data, axis=1)*trigger  # trigger dictates the portion of the maximum ground velocity,
+    thresholds = thresholds.reshape(-1,1)       # which needs to be reached to pick the p-phase
+    data = data / thresholds
+    ind = np.argmax(data>1, axis=1)
+    t = ind*dt
+    return np.transpose(np.vstack((ind,t))) # p-phase output: first column index and second column simulation time
+
+def SWaveDelay(xyz, pphase, hypocenter, centroid):
```

Comment:
you could add a docstring to the function (and check that all other functions have one).

Reply:
Done.